### PR TITLE
Make eraseInfo work for classes with EmptyScopes

### LINKED ISSUE #19506: Missing context bound implicit crashes compiler under `-source:3.4-migration`
New regression found in Open CB for `com-lihaoyi/upickle`
It was not found in the maintained CommunityBuild, because Open CB might automatically add `-source:3.x-migration` flags. 

The issue exists only when we combine `-source:3.4-migration`, method with context bound and at least 1 default method argument.

## Compiler version
Last good release: 3.4.0-RC1-bin-20240109-91db06a-NIGHTLY
First bad release: 3.4.0-RC1-bin-20240112-c50f2ff-NIGHTLY
Bisect points to 13a71ef6075cc8acec3c2cb248251170d7203943

## Minimized code

```Scala
//> using options "-source:3.4-migration",

trait Reader[T]
def read[T: Reader](s: String, trace: Boolean = false): T = ???

@main def Test = read[Object]("") // crash 

```

## Output (click arrow to expand)

<details>

```scala
exception occurred while typechecking /Users/wmazur/projects/dotty/bisect/main.scala

  An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/lampepfl/dotty/issues/new/choose
  For non-enriched exceptions, compile with -Yno-enrich-error-messages.

     while compiling: /Users/wmazur/projects/dotty/bisect/main.scala
        during phase: typer
                mode: Mode(ImplicitsEnabled)
     library version: version 2.13.12
    compiler version: version 3.4.0-RC1-bin-SNAPSHOT-git-13a71ef
            settings: -classpath /Users/wmazur/.ivy2/local/org.scala-lang/scala3-library_3/3.4.0-RC1-bin-SNAPSHOT/jars/scala3-library_3.jar:/Users/wmazur/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.12/scala-library-2.13.12.jar -d /Users/wmazur/projects/dotty/bisect/.scala-build/bisect_1983f628ca-3fe1fc2157/classes/main -java-output-version 17 -source 3.4-migration -sourceroot /Users/wmazur/projects/dotty/bisect

Exception in thread "main" java.util.NoSuchElementException: head of empty list
        at scala.collection.immutable.Nil$.head(List.scala:662)
        at scala.collection.immutable.Nil$.head(List.scala:661)
        at dotty.tools.dotc.typer.Migrations.contextBoundParams(Migrations.scala:115)
        at dotty.tools.dotc.typer.Migrations.contextBoundParams$(Migrations.scala:28)
        at dotty.tools.dotc.typer.Typer.contextBoundParams(Typer.scala:120)
        at dotty.tools.dotc.typer.Typer.adaptToArgs$1(Typer.scala:3776)
        at dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:4360)
        at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:3679)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3275)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3279)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3391)
        at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:956)
        at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1116)
        at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:350)
        at dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:120)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3114)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3198)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3275)
        at dotty.tools.dotc.typer.Typer.addImplicitArgs$1(Typer.scala:3911)
        at dotty.tools.dotc.typer.Typer.adaptNoArgsImplicitMethod$1(Typer.scala:3929)
        at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:4129)
        at dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:4365)
        at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:3679)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3275)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3279)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3391)
        at dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1670)
        at dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1660)
        at dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1670)
        at dotty.tools.dotc.typer.Namer.typedAheadRhs$1$$anonfun$1(Namer.scala:1932)
        at dotty.tools.dotc.inlines.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:256)
        at dotty.tools.dotc.typer.Namer.typedAheadRhs$1(Namer.scala:1932)
        at dotty.tools.dotc.typer.Namer.rhsType$1(Namer.scala:1940)
        at dotty.tools.dotc.typer.Namer.cookedRhsType$1(Namer.scala:1958)
        at dotty.tools.dotc.typer.Namer.lhsType$1(Namer.scala:1959)
        at dotty.tools.dotc.typer.Namer.inferredResultType(Namer.scala:1970)
        at dotty.tools.dotc.typer.Namer.inferredType$1(Namer.scala:1708)
        at dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1714)
        at dotty.tools.dotc.typer.Namer.defDefSig(Namer.scala:1812)
        at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:800)
        at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:943)
        at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:823)
        at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:174)
        at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:189)
        at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:191)
        at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:393)
        at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:3061)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3086)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3197)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3275)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3279)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3301)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3347)
        at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2789)
        at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:3102)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3106)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3197)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3275)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3279)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3301)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3347)
        at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2922)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3148)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3198)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3275)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3279)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3391)
        at dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$1(TyperPhase.scala:47)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at dotty.tools.dotc.core.Phases$Phase.monitor(Phases.scala:477)
        at dotty.tools.dotc.typer.TyperPhase.typeCheck(TyperPhase.scala:53)
        at dotty.tools.dotc.typer.TyperPhase.$anonfun$4(TyperPhase.scala:99)
        at scala.collection.Iterator$$anon$6.hasNext(Iterator.scala:479)
        at scala.collection.Iterator$$anon$9.hasNext(Iterator.scala:583)
        at scala.collection.immutable.List.prependedAll(List.scala:152)
        at scala.collection.immutable.List$.from(List.scala:684)
        at scala.collection.immutable.List$.from(List.scala:681)
        at scala.collection.IterableOps$WithFilter.map(Iterable.scala:898)
        at dotty.tools.dotc.typer.TyperPhase.runOn(TyperPhase.scala:100)
        at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:315)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
        at dotty.tools.dotc.Run.runPhases$1(Run.scala:337)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:348)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:357)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:357)
        at dotty.tools.dotc.Run.compileSources(Run.scala:261)
        at dotty.tools.dotc.Run.compile(Run.scala:246)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
        at dotty.tools.dotc.Driver.process(Driver.scala:197)
        at dotty.tools.dotc.Driver.process(Driver.scala:165)
        at dotty.tools.dotc.Driver.process(Driver.scala:177)
        at dotty.tools.dotc.Driver.main(Driver.scala:207)
        at dotty.tools.dotc.Main.main(Main.scala)
Compilation failed
```
</details>

### LINKED ISSUE #19530: `classOf[scala.Singleton]` compiler crash
## Compiler version

- 3.3.1
- 3.4.0-RC3
- 3.4.1-RC1-bin-20240124-7c3edc5-NIGHTLY

## Minimized code

```Scala
object A {
  def x = classOf[scala.Singleton]
}
```

## Output (click arrow to expand)

<details>

```scala
[info] compiling 1 Scala source to /Users/kenji/classOf-scala-Singleton-crash/target/scala-3.4.0-RC3/classes ...

  exception while retyping classOf[Singleton] of class Literal # -1

  An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/lampepfl/dotty/issues/new/choose
  For non-enriched exceptions, compile with -Yno-enrich-error-messages.

     while compiling: /Users/kenji/classOf-scala-Singleton-crash/A.scala
        during phase: MegaPhase{elimErasedValueType, pureStats, vcElideAllocations, etaReduce, arrayApply, elimPolyFunction, tailrec, completeJavaEnums, mixin, lazyVals, memoize, nonLocalReturns, capturedVars}
                mode: Mode(ImplicitsEnabled)
     library version: version 2.13.12
    compiler version: version 3.4.0-RC3
            settings: -bootclasspath /Users/kenji/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.4.0-RC3/scala3-library_3-3.4.0-RC3.jar:/Users/kenji/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.12/scala-library-2.13.12.jar -classpath /Users/kenji/classOf-scala-Singleton-crash/target/scala-3.4.0-RC3/classes -d /Users/kenji/classOf-scala-Singleton-crash/target/scala-3.4.0-RC3/classes

[error] ## Exception when compiling 1 sources to /Users/kenji/classOf-scala-Singleton-crash/target/scala-3.4.0-RC3/classes
[error] java.lang.UnsupportedOperationException: openForMutations
[error] dotty.tools.package$.unsupported(package.scala:9)
[error] dotty.tools.dotc.core.Scopes$Scope.openForMutations(Scopes.scala:178)
[error] dotty.tools.dotc.core.TypeErasure.dotty$tools$dotc$core$TypeErasure$$apply(TypeErasure.scala:728)
[error] dotty.tools.dotc.core.TypeErasure.eraseInfo(TypeErasure.scala:817)
[error] dotty.tools.dotc.core.TypeErasure$.transformInfo(TypeErasure.scala:273)
[error] dotty.tools.dotc.transform.Erasure.transform(Erasure.scala:96)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:831)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:848)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:877)
[error] dotty.tools.dotc.core.Types$NamedType.computeDenot(Types.scala:2467)
[error] dotty.tools.dotc.core.Types$NamedType.denot(Types.scala:2430)
[error] dotty.tools.dotc.core.Types$NamedType.info(Types.scala:2419)
[error] dotty.tools.dotc.core.Types$Type.isRef(Types.scala:200)
[error] dotty.tools.dotc.ast.tpd$.clsOf(tpd.scala:1345)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedLiteral(Erasure.scala:639)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3117)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3198)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3275)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3279)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3391)
[error] dotty.tools.dotc.typer.Typer.$anonfun$62(Typer.scala:2602)
[error] dotty.tools.dotc.inlines.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:256)
[error] dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:2602)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:959)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3096)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3197)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3275)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3279)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3301)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3347)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1058)
[error] dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2789)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedClassDef(Erasure.scala:1047)
[error] dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:3102)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3106)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3197)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3275)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3279)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3301)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3347)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1058)
[error] dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2922)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3148)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3198)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:174)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3275)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3279)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3391)
[error] dotty.tools.dotc.transform.Erasure.run(Erasure.scala:143)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:354)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:360)
[error] dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:315)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$1(Run.scala:337)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:348)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:357)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:71)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:357)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:261)
[error] dotty.tools.dotc.Run.compile(Run.scala:246)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:141)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:193)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:248)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:183)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:211)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:180)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:178)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:464)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:419)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:506)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:406)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:172)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:425)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2371)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2321)
[error] sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:31)
[error] sbt.internal.io.Retry$.apply(Retry.scala:47)
[error] sbt.internal.io.Retry$.apply(Retry.scala:29)
[error] sbt.internal.io.Retry$.apply(Retry.scala:24)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2319)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:63)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:69)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:283)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
[error] sbt.Execute.work(Execute.scala:292)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:283)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:65)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] java.base/java.lang.Thread.run(Thread.java:829)
```
</details>

### PR DESCRIPTION
**Pull request by odersky**
Fixes #19530

### CODE REVIEW

**Comment by odersky** [comment]
Closed in favor of #19553

File: compiler/src/dotty/tools/dotc/core/Scopes.scala [hunk 1]
```diff
@@ -159,19 +159,27 @@ object Scopes {
...
-        if (!p(sym)) {
-          if (result == null) result = cloneScope
+      for sym <- iterator do
+        def drop() =
+          if result == null then result = cloneScope
           result.nn.unlink(sym)
-        }
+        if keep(sym) then
+          val newName = rename(sym, sym.name)
+          if newName ne sym.name then
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/Scopes.scala:176]: This PR fails probably because `ne` rather than `!=` is used here?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Scopes.scala:176]: I doubt it, but maybe you can try?

Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Scopes.scala:176]: sure, I will take a look tomorrow

Comment (He-Pin) [reply compiler/src/dotty/tools/dotc/core/Scopes.scala:176]: You can edit it directly on phone:)

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Scopes.scala:176]: Indeed. Let me try that.

Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Scopes.scala:176]:
```suggestion
          if newName != sym.name then
```

**Comment by odersky** [comment]
!= vs ne made no difference.

**Comment by noti0na1** [comment]
It seems that using the names of symbols may break the order they are loaded at some places, so I use `null` as the default value when renaming is not needed.

**Comment by odersky** [comment]
LGTM. @noti0na1 If you can approve it, please do. It does not let me do it, since I originated the PR.

**Review by noti0na1 — APPROVED** [review]

**Pull request merged**